src/server/game/Maps/Map.cpp

```diff
@@ -2596,6 +2600,104 @@ ZLiquidStatus Map::getLiquidStatus(float x, float y, float z, uint8 ReqLiquidTyp
+        data.areaInfo.groupId = vmapData.areaInfo.groupId;
+        data.areaInfo.mogpFlags = vmapData.areaInfo.flags;
+    }
+
+
+    GridMap* gmap = const_cast<Map*>(this)->GetGrid(x, y);
+    float mapHeight = gmap->getHeight(x, y);
+
+    // area lookup
+    AreaTableEntry const* areaEntry = nullptr;
+    if (vmapData.hasAreaInfo && !(z + 2.0f > mapHeight && mapHeight > vmapData.floorZ))
+        if (WMOAreaTableEntry const* wmoEntry = GetWMOAreaTableEntryByTripple(vmapData.areaInfo.rootId, vmapData.areaInfo.adtId, vmapData.areaInfo.groupId))
+            areaEntry = sAreaTableStore.LookupEntry(wmoEntry->areaId);
+
+    if (areaEntry)
+        data.areaInfo.areaId = areaEntry->ID;
+    else
+    {
+        if (gmap)
+            data.areaInfo.areaId = gmap->getArea(x, y);
+        else
+            data.areaInfo.areaId = 0;
+
+        if (!data.areaInfo.areaId)
+            data.areaInfo.areaId = i_mapEntry->linked_zone;
+
+        if (data.areaInfo.areaId)
+            areaEntry = sAreaTableStore.LookupEntry(data.areaInfo.areaId);
+    }
+
+    // liquid processing
+    data.liquidStatus = LIQUID_MAP_NO_WATER;
+    if (vmapData.hasLiquid && vmapData.liquidInfo.level > vmapData.floorZ && z + 2.0f > vmapData.floorZ)
+    {
+        uint32& liquidType = vmapData.liquidInfo.type;
+        if (GetId() == 530 && vmapData.liquidInfo.type == 2) // gotta love blizzard hacks
+            liquidType = 15;
+
+        uint32 liquidFlagType = 0;
+        if (LiquidTypeEntry const* liquidData = sLiquidTypeStore.LookupEntry(liquidType))
+            liquidFlagType = liquidData->Type;
+
+        if (liquidType && liquidType < 21 && areaEntry)
+        {
+            uint32 overrideLiquid = areaEntry->LiquidTypeOverride[liquidFlagType];
+            if (!overrideLiquid && areaEntry->zone)
+            {
+                AreaTableEntry const* zoneEntry = sAreaTableStore.LookupEntry(areaEntry->zone);
+                if (zoneEntry)
+                    overrideLiquid = zoneEntry->LiquidTypeOverride[liquidFlagType];
+            }
+
+            if (LiquidTypeEntry const* overrideData = sLiquidTypeStore.LookupEntry(overrideLiquid))
+            {
+                liquidType = overrideLiquid;
+                liquidFlagType = overrideData->Type;
+            }
+        }
+
+        data.liquidInfo.level = vmapData.liquidInfo.level;
+        data.liquidInfo.depth_level = vmapData.floorZ;
+        data.liquidInfo.entry = liquidType;
+        data.liquidInfo.type_flags = 1 << liquidFlagType;
+
+        float delta = vmapData.liquidInfo.level - z;
+        if (delta > 2.0f)
+            data.liquidStatus = LIQUID_MAP_UNDER_WATER;
+        else if (delta > 0.0f)
+            data.liquidStatus = LIQUID_MAP_IN_WATER;
+        else if (delta > 0.1f)
+            data.liquidStatus = LIQUID_MAP_WATER_WALK;
+        else
+            data.liquidStatus = LIQUID_MAP_ABOVE_WATER;
+    }
+    // look up liquid data from grid map
+    if (gmap)
```

Comment:
broken. you are overwriting vmap result. `Map::getLiquidStatus` returns early if vmap result is found